bin/prover-client/src/operators/cl_stf.rs

```diff
@@ -49,56 +49,76 @@ impl ClStfOperator {
         }
     }
 
-    /// Retrieves the [`L2BlockId`] for the given `block_num`
-    pub async fn get_id(&self, block_num: u64) -> Result<L2BlockId, ProvingTaskError> {
-        let l2_headers = self
+    pub async fn get_exec_id(&self, cl_block_id: L2BlockId) -> Result<Buf32, ProvingTaskError> {
+        let header = self
             .cl_client
-            .get_headers_at_idx(block_num)
+            .get_header_by_id(cl_block_id)
             .await
-            .inspect_err(|_| error!(%block_num, "Failed to fetch l2_headers"))
-            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?;
+            .inspect_err(|_| error!(%cl_block_id, "Failed to fetch corresponding ee data"))
+            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?
+            .ok_or_else(|| {
+                error!(%cl_block_id, "Block not found");
+                ProvingTaskError::InvalidWitness("Invalid block".to_string())
+            })?;
 
-        let cl_stf_id_buf: Buf32 = l2_headers
-            .expect("invalid height")
-            .first()
-            .expect("at least one l2 blockid")
-            .block_id
-            .into();
-        Ok(cl_stf_id_buf.into())
+        let block = self.evm_ee_operator.get_block(header.block_idx).await?;
+        Ok(block.header.hash.into())
     }
 
-    /// Retrieves the slot num of the given [`L2BlockId`]
-    pub async fn get_slot(&self, id: L2BlockId) -> Result<u64, ProvingTaskError> {
-        let header = self
+    /// Retrieves the previous [`L2BlockId`] for the given `L2BlockId`
+    pub async fn get_prev_block_id(
+        &self,
+        block_id: L2BlockId,
+    ) -> Result<L2BlockId, ProvingTaskError> {
+        let l2_block = self
             .cl_client
-            .get_header_by_id(id)
+            .get_header_by_id(block_id)
             .await
-            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?
-            .expect("invalid blkid");
-        Ok(header.block_idx)
+            .inspect_err(|_| error!(%block_id, "Failed to fetch l2_header"))
+            .map_err(|e| ProvingTaskError::RpcError(e.to_string()))?;
+
+        let prev_block: Buf32 = l2_block
+            .ok_or_else(|| {
+                error!(%block_id, "Block not found");
```

Comment:
Couple of things:

* it's probably be better to have a method like `get_block_ancestors(id: L2BlockId, n_ancestors: u32) -> Result<Vec<L2BlockId>, _>`, which means we don't have to repeatedly call a "get parent" method to walk back over the chain, and it aligns slightly better with the use case since we know how long we *expect* the batches we're proving to be and can validate the block we expect it to be later
* The logging strings here are rather vague and imprecise.  Seeing "block not found" in the log message without context isn't helpful.  Also should be consistent about `block_id`, `blockid`, `blkid`, `l2blkid`, etc.
* Also not really comfortable with the RPC error variants being sorta handwaved like this.  Not sure what an ideal solution here is though.